SVGLoader: Implement custom createShapes() method.

Ttommeke opened the pull request:
Related issue: Fixed #16950.

**Description**

The current implementation of `ShapePath` only allowed the `toShapes()` function to generate shapes with holes correctly if the path orientation(`CW `or `CCW`) was known before hand. This made it difficult to import and convert a user generated SVG via the `SVGLoader` into a `Shape` object.

This implementation adds a scanline that will automatically determine the outside of a shape and its orientation. By passing the `fillRule` argument (same as the SVG style `fill-rule`), it can proceed to generate a correct shape.

**Some issues that need to be addressed:**
- This solution can't yet transform self intersecting paths correctly like the star shape in the documentation. [WindingRule](https://svgwg.org/svg2-draft/painting.html#WindingRule)
- It breaks the old implementation because It got rid of the `isCCW` and `noHoles` parameter and replaced them with `fillRule`.
- I have not written unit tests... Am I going to burn on the stake now?

Ttommeke committed:
added a scanline method to add support
for svg fillrule

Ttommeke committed:
changed ShapePath

Ttommeke commented:
I see that i still have to do some linting.

Ttommeke committed:
lint pass

Mugen87 commented:
> It breaks the old implementation because It got rid of the isCCW and noHoles parameter and replaced them with fillRule.

I don't think this PR can be accepted then.

Would it be possible to move this implementation to `SVGLoader` so the loader does not rely on `ShapePath` anymore?

Mugen87 commented on package.json:
```diff
@@ -89,7 +89,7 @@
     "@rollup/plugin-node-resolve": "^11.2.0",
     "concurrently": "^5.3.0",
-    "eslint": "^7.20.0",
+    "eslint": "^7.21.0",
```
Please remove all changes to `package.json` and `package-lock.json`.

Ttommeke replied:
Done! I however have no idea why he flags a change in the last rule of `package-lock.json`.

Ttommeke committed:
lint pass

Ttommeke committed:
Merge branch 'issue/ShapePath-toShapes-scanlineMethod' of github.com:Ttommeke/three.js into issue/ShapePath-toShapes-scanlineMethod

Ttommeke committed:
removed edits in package.json

Ttommeke commented:
> > It breaks the old implementation because It got rid of the isCCW and noHoles parameter and replaced them with fillRule.
> 
> I don't think this PR can be accepted then. A lot of user level code will going to break.
> 
> Would it be possible to move this implementation to `SVGLoader` so the loader does not rely on `ShapePath` anymore?

Thanks for the feedback.
I understand. I'm not sure if it is possible to remove `ShapePath` from the implementation, but i could create a `createShape(ShapePath)` on `SVGLoader` instead.

Mugen87 commented:
> but i could create a createShape(ShapePath) on SVGLoader instead.

That sounds like a plan 👍 .

mrdoob commented:
> > but i could create a createShape(ShapePath) on SVGLoader instead.
>
> That sounds like a plan 👍 .

Yep, sounds much more safe.

Ttommeke committed:
fix refectoring

Ttommeke committed:
removing useless imports.

mrdoob commented on examples/jsm/loaders/SVGLoader.js:
```diff
@@ -1680,6 +1684,463 @@ SVGLoader.pointsToStroke = function ( points, style, arcDivisions, minDistance )
+
+		}
+
+		if ( sa < - 0.0000000001 ) {
```
You can probably use `Number.EPSILON` here?

Ttommeke replied:
I'll try that. Do you also have such a Number for my `BIGNUMBER` constant? I tried `Number.MAX_SAFE_INTEGER` but then the algorithm acts up.

yomboprime commented:
Hi, I've done some tests. I hate to bring bad news, here are some results. I will look with more insight in some days or when you make more commits.

For the tests I've made the following:
```
// const shapes = path.toShapes( true );
const shapes = SVGLoader.createShapes( path );
```

Results:
- The tiger gives error 'baseIntersections[0] is undefined'.
- The file 'Units' shows some letter holes which are obscured:

![imagen](https://user-images.githubusercontent.com/2453652/110163524-6c728680-7df0-11eb-9330-8ce82dbab722.png)
![imagen](https://user-images.githubusercontent.com/2453652/110163589-857b3780-7df0-11eb-993a-88d59d336f0f.png)

(BTW, you forgot to modify the js and not the jsm)

yomboprime commented:
With the following changes the tiger shows up. I've not managed to make the text holes look correct.

Added this ```if``` before this line:
```
if ( baseIntersections.length === 0 ) return { identifier: simplePath.identifier, isHole: false, for: null };

const firstXOfPath = baseIntersections[ 0 ].point.x;
```

And added this check in this ```if```:
```
if ( ! amIAHole.isHole && p.points.length > 0 ) {

	const shape = new Shape( p.points );
```

Ttommeke commented:
@yomboprime That is a good fix. I'll sort out the text holes. Shouldn't be that difficult.

Ttommeke commented:
> (BTW, you forgot to modify the js and not the jsm)
I feel a bit stupid for asking, but i'm not sure what i have to do now. Do i have to edit both? Or only one?

Mugen87 commented:
You always edit files in `exampels/js` and then use `node utils/modualrize.js` to generate the ES6 module file.

yomboprime commented:
BTW, I think all the polarAngle calculations can be removed, since the theta angle is not used anywhere?

Ttommeke committed:
solving hole issue when scanline goes through point

Ttommeke committed:
finish

Ttommeke commented:
- That angle is indeed unnecessary. I Removed it.
- The `0.0000001` is replaced with `Number.Epsilon`.
- I removed beforehand any empthy paths to fix the `baseintersection[0]` is undefined.
- The undetected holes where caused by intersections through points. I fixed the issue by merging intersections with a `t` value within a margin of `Number.EPSILON`.

yomboprime commented:
Please, could you add this SVG file to the example in ```test/testDefs/```? (renaming to .svg)
[defs5.txt](https://github.com/mrdoob/three.js/files/6098095/defs5.txt)


Closes #16950 #19168

Mugen87 commented on package-lock.json:
```diff
@@ -3688,4 +3688,4 @@
     }
   }
-}
+}
```
Do you mind undo this change? There should be no diff in `package-lock.json`.

Ttommeke replied:
Second time... Fixed it now. Sorry!

Mugen87 commented on examples/js/loaders/SVGLoader.js:
```diff
@@ -1614,6 +1614,431 @@ THREE.SVGLoader.prototype = Object.assign( Object.create( THREE.Loader.prototype
+				} else if ( classify.loc == IntersectionLocationType.BETWEEN ) {
+
+					x = + ( ( x1 + classify.t * ( x2 - x1 ) ).toPrecision( 10 ) );
+					y = + ( ( y1 + classify.t * ( y2 - y1 ) ).toPrecision( 10 ) );
```
It seems the variables `x` and `y` are not declared before use.

Ttommeke replied:
True. Thanks, fixed that!

Ttommeke commented:
@yomboprime Yes i can.

yomboprime commented:
I've modified the functions ```findEdgeIntersection```, ```classifyPoint``` and ```getIntersections``` for some speed and memory optimization. Basically the parameters and return values being passed. Please tell me what you think.

Here they are in a file:

[modifiedFunctions.txt](https://github.com/mrdoob/three.js/files/6126154/modifiedFunctions.txt)

And reproduced here:
```javascript
	const classifyResult = {
		loc: IntersectionLocationType.ORIGIN,
		t: 0
	}

	function findEdgeIntersection( a0, a1, b0, b1 ) {

		var x1 = a0.x;
		var x2 = a1.x;
		var x3 = b0.x;
		var x4 = b1.x;
		var y1 = a0.y;
		var y2 = a1.y;
		var y3 = b0.y;
		var y4 = b1.y;
		var nom1 = ( x4 - x3 ) * ( y1 - y3 ) - ( y4 - y3 ) * ( x1 - x3 );
		var nom2 = ( x2 - x1 ) * ( y1 - y3 ) - ( y2 - y1 ) * ( x1 - x3 );
		var denom = ( y4 - y3 ) * ( x2 - x1 ) - ( x4 - x3 ) * ( y2 - y1 );
		var t1 = nom1 / denom;
		var t2 = nom2 / denom;

		if ( ( ( denom === 0 ) && ( nom1 !== 0 ) ) || ( t1 <= 0 ) || ( t1 >= 1 ) || ( t2 < 0 ) || ( t2 > 1 ) ) {

			//1. lines are parallel or edges don't intersect

			return null;

		} else if ( ( nom1 === 0 ) && ( denom === 0 ) ) {

			//2. lines are colinear

			//check if endpoints of edge2 (b0-b1) lies on edge1 (a0-a1)
			for ( var i = 0; i < 2; i ++ ) {

				classifyPoint( i === 0 ? b0 : b1, a0, a1 );
				//find position of this endpoints relatively to edge1
				if ( classifyResult.loc == IntersectionLocationType.ORIGIN ) {

					var point = ( i === 0 ? b0 : b1 );
					return { x: point.x, y: point.y, t: classifyResult.t };

				} else if ( classifyResult.loc == IntersectionLocationType.BETWEEN ) {

					x = + ( ( x1 + classifyResult.t * ( x2 - x1 ) ).toPrecision( 10 ) );
					y = + ( ( y1 + classifyResult.t * ( y2 - y1 ) ).toPrecision( 10 ) );
					return { x: x, y: y, t: classifyResult.t, };

				}

			}

			return null;

		} else {

			//3. edges intersect

			for ( var i = 0; i < 2; i ++ ) {

				classifyPoint( i === 0 ? b0 : b1, a0, a1 );

				if ( classifyResult.loc == IntersectionLocationType.ORIGIN ) {

					var point = ( i === 0 ? b0 : b1 );
					return { x: point.x, y: point.y, t: classifyResult.t };

				}

			}

			var x = + ( ( x1 + t1 * ( x2 - x1 ) ).toPrecision( 10 ) );
			var y = + ( ( y1 + t1 * ( y2 - y1 ) ).toPrecision( 10 ) );
			return { x: x, y: y, t: t1 };

		}

	}

	function classifyPoint( p, edgeStart, edgeEnd ) {

		var ax = edgeEnd.x - edgeStart.x;
		var ay = edgeEnd.y - edgeStart.y;
		var bx = p.x - edgeStart.x;
		var by = p.y - edgeStart.y;
		var sa = ax * by - bx * ay;

		if ( ( p.x === edgeStart.x ) && ( p.y === edgeStart.y ) ) {

			classifyResult.loc = IntersectionLocationType.ORIGIN;
			classifyResult.t = 0;
			return;

		}

		if ( ( p.x === edgeEnd.x ) && ( p.y === edgeEnd.y ) ) {

			classifyResult.loc = IntersectionLocationType.DESTINATION;
			classifyResult.t = 1;
			return;

		}

		if ( sa < - Number.EPSILON ) {

			classifyResult.loc = IntersectionLocationType.LEFT;
			return;

		}

		if ( sa > Number.EPSILON ) {

			classifyResult.loc = IntersectionLocationType.RIGHT;
			return;


		}

		if ( ( ( ax * bx ) < 0 ) || ( ( ay * by ) < 0 ) ) {

			classifyResult.loc = IntersectionLocationType.BEHIND;
			return;

		}

		if ( ( Math.sqrt( ax * ax + ay * ay ) ) < ( Math.sqrt( bx * bx + by * by ) ) ) {

			classifyResult.loc = IntersectionLocationType.BEYOND;
			return;

		}

		var t;

		if ( ax !== 0 ) {

			t = bx / ax;

		} else {

			t = by / ay;

		}

		classifyResult.loc = IntersectionLocationType.BETWEEN;
		classifyResult.t = t;

	}

	function getIntersections( path1, path2 ) {

		const intersectionsRaw = [];
		const intersections = [];

		for ( let index = 1; index < path1.length; index ++ ) {

			const path1EdgeStart = path1[ index - 1 ];
			const path1EdgeEnd = path1[ index ];

			for ( let index2 = 1; index2 < path2.length; index2 ++ ) {

				const path2EdgeStart = path2[ index2 - 1 ];
				const path2EdgeEnd = path2[ index2 ];

				const intersection = findEdgeIntersection( path1EdgeStart, path1EdgeEnd, path2EdgeStart, path2EdgeEnd );

				if ( intersection !== null && intersectionsRaw.find(i => i.t <= intersection.t + Number.EPSILON && i.t >= intersection.t - Number.EPSILON) === undefined ) {

					intersectionsRaw.push( intersection );
					intersections.push( new THREE.Vector2( intersection.x, intersection.y ) );

				}

			}

		}

		return intersections;

	}
```

Ttommeke committed:
added yomboprime's function revisions and his text example

Ttommeke committed:
fix package-lock

Ttommeke committed:
fix unintialized variables

Ttommeke commented:
Sorry for the wait.

@yomboprime That looks great. I've added to the code. I've a question out of curiosity. Why did you choose to not return the classification of an intersection and instead use a variable(`classifyResult`) that you overwrite every time. Is it faster that way?

I believe to have fixed all of the issues @Mugen87 found in my pull request.

Mugen87 commented on examples/jsm/loaders/SVGLoader.js:
```diff
@@ -1626,6 +1629,443 @@ SVGLoader.prototype = Object.assign( Object.create( Loader.prototype ), {
+				} else if ( classifyResult.loc == IntersectionLocationType.BETWEEN ) {
+
+					x = + ( ( x1 + classifyResult.t * ( x2 - x1 ) ).toPrecision( 10 ) );
+					y = + ( ( y1 + classifyResult.t * ( y2 - y1 ) ).toPrecision( 10 ) );
```
It seems you have to generate the module one more time.

Ttommeke replied:
Yes!

Ttommeke replied:
Done.

Ttommeke committed:
re modularize

yomboprime commented:
>  I've a question out of curiosity. Why did you choose to not return the classification of an intersection and instead use a variable(`classifyResult`) that you overwrite every time. Is it faster that way?

Yes, should be faster as the JS engine doesn't have to create an additional object to return it and then just forget it after use. The V8 probably does a good job with such return-and-forget objects but it still should be faster. Since the intersection call is deeply nested in paths and segments loops, the optimization is multiplied potentially thousands of times.

Mugen87 commented:
https://raw.githack.com/Ttommeke/three.js/issue/ShapePath-toShapes-scanlineMethod/examples/index.html?q=svg#webgl_loader_svg

Mugen87 commented:
@yomboprime Do you know why the test files `Defs` and `Defs2` are not working?

Mugen87 commented:
@Ttommeke Do you mind changing this line in the editor, too?

https://github.com/mrdoob/three.js/blob/5635d4b39c42e5e624954562e84878a92fe1bc57/editor/js/Loader.js#L495

yomboprime commented:
> @yomboprime Do you know why the test files `Defs` and `Defs2` are not working?

I noticed it some time ago and was going to look at it when I have some spare time.

Mugen87 commented:
Okay. Just wanted to make sure this is unrelated to the PR 👍

Mugen87 commented:
@yomboprime Found the issue, see https://github.com/mrdoob/three.js/commit/330b18605ae145604f4168cf70802621f11ea2a1#r48264467.

Ttommeke commented:
I was writing the documentation when i realized that most other loaders do not define static methods. They also all implement a parse method. For the sake of keeping things organized and the same,  would it be a good idea to refactor the function createShapes(as a static method) to a parse(non static) method?

yomboprime commented:
> I was writing the documentation when i realized that most other loaders do not define static methods. They also all implement a parse method. For the sake of keeping things organized and the same, would it be a good idea to refactor the function createShapes(as a static method) to a parse(non static) method?

I don't understand... SVGLoader already has a ```parse``` method.

Ttommeke commented:
You are right. I was wrong. I shouldn't be thinking about refactoring code at 1 am.

mrdoob commented:
Is this ready to merge then?

Mugen87 commented:
It seems https://github.com/mrdoob/three.js/pull/21380#issuecomment-799285497 is still open.

yomboprime commented:
I think yes. But Tom was doing documentation, I understood.

Ttommeke commented:
Yes. I'm finishing up now.

Ttommeke committed:
documentation done

Ttommeke commented:
Believe it's ready.

mrdoob commented:
Alright!

mrdoob commented:
Thanks! 🙏

Ttommeke commented:
It was a pleasure! Thanks for the great experience! @yomboprime @Mugen87 @mrdoob 

First public PR ever.
